modules/turnstile/turnstile.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package turnstile
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"code.gitea.io/gitea/modules/json"
+	"code.gitea.io/gitea/modules/setting"
+)
+
+// Response is the structure of JSON returned from API
+type Response struct {
+	Success     bool        `json:"success"`
+	ChallengeTS string      `json:"challenge_ts"`
+	Hostname    string      `json:"hostname"`
+	ErrorCodes  []ErrorCode `json:"error-codes"`
+	Action      string      `json:"login"`
+	Cdata       string      `json:"cdata"`
+}
+
+// Verify calls Cloudflare Turnstile API to verify token
+func Verify(ctx context.Context, response, ip string) (bool, error) {
+	// Cloudflare turnstile official access instruction address: https://developers.cloudflare.com/turnstile/get-started/server-side-validation/
+	post := url.Values{
+		"secret":   {setting.Service.CfTurnstileSecret},
+		"response": {response},
+		"remoteip": {ip},
```

Comment:
The `remoteip` is optional. Does Gitea guarantee that the site admin always has the correct reverse-proxy config and Gitea can get the correct user remote IP by RemoteAddr? If no, what would happen? Just a question.

Reply:
My personal opinion is the same as this [answer](https://security.stackexchange.com/questions/81865/is-there-any-reason-to-include-the-remote-ip-when-using-recaptcha/81951#81951). Therefore, I think it may be a better choice to carry `remoteip` in the parameter.
> One possible scenario is farming cheap labour to manually solve captchas and then submit them back with the form.

---

But the usage of `remoteip` above is just my personal guess, cloudflare’s official documentation does not explain how they use `remoteip`.
If you think that third parties should not be given so much information, then I will remove `remoteip` in a future commit.
What do you think?

Reply:
Personally I prefer to  keep the same behavior as hcaptcha (no `remoteip`)

It might still be fine to have the `remoteip` for the cloudflare turnstile, then it could be better to have enough comments.

Reply:
Ok, I'm going to remove `remoteip`. If there is enough information in the future to prove that it would be better to carry this parameter, then we will discuss whether to add it.

Reply:
It was removed in this commit https://github.com/go-gitea/gitea/pull/22369/commits/46aa739879cd5a199a3d74be43b1a66c6e954a05